proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+m ::= {..., tags tagaddr*}
+```
+
+#### Administrative Instructions
+
+```
+instr ::= ... | throw tagaddr | catch_n{ tagaddr? instr* }* instr* end
+        | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end
+```
+
+#### Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts.
+
+```
+B^0 ::= val* [_] instr*
+B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end
+      | delegate{ labelidx } B^k end
+B^{k+1} ::= val* (label_n{instr*} B^k end) instr*
+```
+
+#### Throw Contexts
+
+Throw contexts don't skip over handlers, they are used to match a thrown exception with the innermost handler.
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{ tagaddr val^n } T end
+   | frame_n{F} T end
+```
+
+Note that because `catch_n` instructions are not included above, there is always a unique maximal throw context.
+
+### Reduction of instructions
+
+Reduction steps for the new instructions or administrative instructions.
+
+```
+F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+An absent tag address in a `handler` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{ε instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l} val^n instr* end) end
+  (if bt = [t^n]→[t^m])
+
+delegate{l} val^n end ↪ val^n
+
+label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end
+  ↪ val^n (throw a)
+```
+
+Note that the last reduction step above is similar to the reduction of `br`.
+
+### Typing rules for administrative instructions
+
+```
+S.tags[a].type = [t*]→[]
+--------------------------------
+S;C ⊢ throw a : [t1* t*]→[t2*]
+
+((S.tags[a].type = [t'*]→[])?
+ S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
+S;C, labels [t*] ⊢ instr* : []→[t*]
+-----------------------------------------------------------------
+S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]
+
+S;C, labels [t*] ⊢ instr* : []→[t*]
+C.labels[l] = [t'*]
```

Comment:
I believe this should actually be
```suggestion
C.labels[l+1] = [t'*]
```
Otherwise, it would be referring to a different label than the original try-delegate.

Reply:
@rossberg: I think the +1 is missing because of the behaviour of "breaking inside the label". This way we also cover the case of delegating to the function body. WDYT?

Reply:
Are you sure? If you have
```
try ... delegate $l
```
you'll reduce to
```
label_m{} (delegate{l} val^n instr* end) end
```
When you type this, the outer `label` adds an entry to the labels environment before you get to this rule, unlike with the `try-delegate` before the reduction. So you have to skip over that.

Reply:
The problem I see is that (4) is not the premise you'll need when arriving at the adm.delegate rule. Instead you'll need
```
(4') C'[l'] = [t'*]
```
with `C' = C, labels [t^n]`, because that's the context you now enter the rule with. The only suitable `l'` that allows you to derive this from (4) is `l+1`, I believe.

Reply:
If we change the current rule
```
F; val^n (try bt instr* delegate l)
  ↪ F; label_m{} (delegate{l} val^n instr* end) end
```
to
```
F; val^n (try bt instr* delegate l)
  ↪ F; delegate(l) (label_m{} val^n instr* end) end
```
meaning, if we put the label created by the try-delegate itself within the scope of `adm.delegate`, can this confusion be helped?

Reply:
Continuing the discussion with a suggestion in #205 .